docs/content/sips/000-app-config.md

```diff
@@ -0,0 +1,143 @@
+title = "SIP xxx - Application Configuration"
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: A configuration system for Spin applications.
+
+Owner: lann.martin@fermyon.com
+
+Created: March 22, 2022
+
+Updated: April 1, 2022
+
+## Background
+
+It is common for applications to require configuration at runtime that isn't known at build time or is too sensitive to be stored with build artifacts:
+
+- Logging configuration
+- Per-channel (production, staging, etc) service dependency URLs
+- Database secrets
+
+## Proposal
+
+### Configuration is defined by components and applications
+
+Configuration within a "parent" (component or application) consists of a number of configuration "slots":
+- Slots are uniquely identified within their parent by a string "key"; in order to allow unambiguous conversion to environment variables or file paths, keys are constrained:
+  - Keys must start with a letter (required for env vars)
+  - Keys consist of only lowercase ascii alphanum and `_` (`[a-z0-9_]`)
+    - Only one `_` at a time and not at the end (to allow delimiting in env vars with `__`)
+- A slot must _either_ be marked as "required" _or_ must be given a default value
+- A slot may be marked as "secret", in which case any associated value should be handled with care (e.g. not logged)
+
+```toml
+[config]
+# This simple form...
+key1 = "default_value"
+# ...is equivalent to:
+key1 = { default = "default_value" }
+# required & secret slot 
+key2 = { required = true, secret = true }
+```
+
+Default values can use template strings to reference other slots.
+```toml
+[config]
+key1 = { required = true }
+key2 = "prefix-{{ key1 }}-suffix"
```

Comment:
What happens if there is a circular reference?  Do we assume that something will have a value injected to break the cycle, and then derive the remainder from there?  (And error if everything is left at default.)  This feels like it could get complicated, identifying which direction to propagate depending on which value is set.  Would it be safer to mandate that the graph of default value references must be acyclic?

Reply:
Yes, I should specify that circular references are invalid. The prototype code enforces this with a simple recursion depth limit.